modules/memfault-firmware-sdk/memfault_gnss_metrics.c

```diff
@@ -0,0 +1,141 @@
+// be notified when a fix request is made. Therefore, we wrap the functions of interest:
+//        - location_init() -- registers an event callback. Save the application
+//              callback and register a wrapper callback.
+//        - location_request() -- initiates a request. Intercept it's result, and
+//              start a session if the request is successful.
+int __real_location_init(location_event_handler_t event_handler);
+int __wrap_location_init(location_event_handler_t event_handler)
+{
+	real_location_event_handler = event_handler;
+	return __real_location_init(memfault_location_event_handler);
+}
+
+// Ensure the substituted function signature matches the original function
+_Static_assert(__builtin_types_compatible_p(__typeof__(&location_init),
+					    __typeof__(&__wrap_location_init)) &&
+		       __builtin_types_compatible_p(__typeof__(&location_init),
+						    __typeof__(&__real_location_init)),
+	       "Error: Wrapped function does not match original location_init function signature");
+
+// Also wrap the location request to know when to start a session
+int __real_location_request(const struct location_config *config);
+int __wrap_location_request(const struct location_config *config)
+{
+	int result = __real_location_request(config);
+
+	if (result == 0) {
+		memfault_gnss_metrics_start_fix_session();
+	}
+	return result;
+}
+
+// Ensure the substituted function signature matches the original function
+_Static_assert(
+	__builtin_types_compatible_p(__typeof__(&location_request),
+				     __typeof__(&__wrap_location_request)) &&
+		__builtin_types_compatible_p(__typeof__(&location_request),
+					     __typeof__(&__real_location_request)),
+	"Error: Wrapped function does not match original location_request function signature");
+
+static void memfault_location_event_handler(const struct location_event_data *event_data)
+{
+	if (event_data->id == LOCATION_EVT_LOCATION || event_data->id == LOCATION_EVT_TIMEOUT ||
+	    event_data->id == LOCATION_EVT_ERROR || event_data->id == LOCATION_EVT_RESULT_UNKNOWN) {
+		memfault_gnss_metrics_stop_fix_session(event_data);
+	}
+	real_location_event_handler(event_data);
+}
+
+/**
+ * @brief Start a GNSS fix session.
+ *
+ */
+static void memfault_gnss_metrics_start_fix_session(void)
+{
+	if (session_in_progress) {
+		return;
+	}
+
+	session_in_progress = true;
+	LOG_DBG("Starting GNSS session");
+	MEMFAULT_METRICS_SESSION_START(ncs_gnss);
+	MEMFAULT_METRIC_ADD(ncs_gnss_fix_request_count, 1);
+}
+
+/**
+ * @brief Stop a GNSS fix session.
+ *
+ * @param event_data pointer to location event data
+ */
+static void memfault_gnss_metrics_stop_fix_session(const struct location_event_data *event_data)
+{
+	if (!session_in_progress) {
+		return;
+	}
+
+	uint32_t session_time_ms;
+	memfault_metrics_heartbeat_timer_read(
+		MEMFAULT_METRICS_KEY(MEMFAULT_METRICS_SESSION_TIMER_NAME(ncs_gnss)),
+		&session_time_ms);
```

Comment:
We technically mark this function as for debug only. Instead we could use `memfault_platform_get_time_since_boot_ms` (which just wraps `k_uptime_get`) to get the uptime at the start of the session. Then when an event happens again get the uptime and calculate the difference and use that to set the metric value instead

Reply:
Ah of course, good idea!